challenges/cloud/sre.md

```diff
@@ -0,0 +1,376 @@
+time and demonstrate that you've spent time thinking about this problem and
+provide a clear path to a solution.
+
+Consider making other reasonable trade-offs. Make sure you communicate them to
+the interview team.
+
+Here are some other trade-offs that will help you to spend less time on the task:
+
+* Do not implement a system that scales or is highly performing. Describe which
+  performance improvements you would add in the future.
+* It is OK if the system is not highly available. Write down how you would make
+  the system highly-available and why your system is not.
+
+## Pitfalls and Gotchas
+
+To help you out, we've composed a list of things that previously resulted in a no-pass from the interview team:
+
+* Scope creep. Candidates have tried to implement too much and ran out of time
+  and energy. To avoid this pitfall, use the simplest solution that will work.
+  Avoid writing too much code. We've seen candidates' code introducing caching
+  and making many mistakes in the caching layer validation logic. Not having
+  caching would have solved this problem.
+  
+* Overly complex designs. Keep things simple and try and eliminate as many moving
+  parts as possible. This is not only going to help in reviewing the solution,
+  but is also often a way to distill a design to its essential parts.
+  
+* Unstructured code. We've seen candidates leaving commented chunks of code,
+  having one large file with all the code, not having code structure at all.
+  
+* Not communicating. Some candidates have submitted all their code to the master
+  branch without raising pull requests, which does not give us the ability to
+  provide feedback on the various implementation phases. We are a distributed
+  team, so structured, asynchronous communication is critical to us.
+
+## Questions
+ 
+It is OK to ask the interview team questions. Some folks stay away from asking
+questions to avoid appearing less experienced, so we provide examples of
+questions to ask and questions we expect candidates to figure out on their own.
+
+Here is a great question to ask:
+
+> Is it OK to assume there will be only a single target kubernetes cluster for
+  this service? I will add a note on how support for multiple clusters could be
+  implemented, but it feels like an unnecessary complexity.
+
+This is the question we expect candidates to figure out on their own:
+
+> What version of Go should I use? What dependency manager should I use? What
+  framework/tool should I use to automate testing and deployment ?
+
+Unless specified in the requirements, pick the solution that works best for you.
+
+# Tools
+
+This task should be implemented in Go and should work on a 64-bit Linux or MacOS
+machine.
+
+It's safe to assume a working Docker environment will be available locally as well.
+
+Additional external dependencies are acceptable such as Terraform and minikube/k3d,
+but please ensure your Makefile is handles these dependencies.
```

Comment:
```suggestion
Additional external dependencies are acceptable such as Terraform and minikube/k3d,
but please ensure your Makefile has targets to install these dependencies.
```

Reply:
Seems fine to just have the makefile spit out "terraform: command not found" as long as instructions for how to install it are linked, don't you think?

Make is a bit esoteric and writing makefiles to install tools that could easily be installed via brew (for example) doesn't seem the best use of a candidate's time.

Reply:
That's fair.  This change was motivated more out of a desitre be clear when we say the `Makefile` should "handle" the dependencies.  Im happy with either an installation target or context aware failures as long as it is clear for the candidate what is expected.